daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/api/Layer.java

```diff
@@ -0,0 +1,91 @@
+ * so that they can be found and dynamically loaded.
+ * <p/>
+ * The SPI creates an instance the class of which is used as a factory to create the
+ * instances actually used by the Daffodil runtime. Compilation of the static information about
+ * the layer occurs only once and is then shared by all runtime instances.
+ * <p/>
+ * Instances of derived layer classes can be stateful. They are private to threads, and each time a layer
+ * is encountered during parse/unparse, an instance is created for that situation.
+ * <p/>
+ * Layer instances should not share mutable state (such as via singleton objects)
+ * <p/>
+ * All the static information about the layer is provided in the arguments.
+ * <p/>
+ * The rest of the Layer class implements the
+ * layer decode/encode logic, which is done as part of deriving one's Layer class from the
+ * Layer base class.
+ * <p/>
+ * About variables: Layer logic may read and write variables. Variables being read are parameters to
+ * the layer algorithm. Variables being written are outputs (such as checksums) from the layer algorithm.
+ * Variables being written must be undefined, since variables in DFDL are single-assignment.
+ * Variables being read must be defined before being read by the layer, and this is true for both
+ * parsing and unparsing. When unparsing, variables being read cannot be forward-referencing to parts
+ * of the DFDL infoset that have not yet been unparsed.
+ * <p/>
+ */
+public abstract class Layer {
+
+  protected final String layerName;
+
+  public Layer(String layerName) {
+    this.layerName = layerName;
+  }
+
+  public final String name() {
+    return this.layerName; // name() method with empty args is required by SPI loader
+  }
+
+  /**
+   * Called exactly once when the schema is compiled to do extra checking that the layer is being used properly.
+   * The thrown exception becomes a SchemaDefinitionError at schema compile time.
+   *
+   * Example checks are:
+   * - layerEncoding is constant and is a single-byte charset
+   * - layerLength, if constant, is within a maximum value range
+   * - layerBoundaryMark string, if constant, is not too long and contains only allowed characters.
+   * These things can be required to be constant by this check, or it can check their values for legality
+   * if they happen to be constant. Since these are runtime-valued properties (can be expressions), then if the
+   * layer allowed that, they must also be checked at runtime.
+   *
+   * You don't have to check that the variables are defined and declared in matching manner, that happens automatically.
+   *
+   * @throws LayerException on any error.
+   *
+   */
+  public void check(LayerCompileInfo layerPropertyInfo) throws LayerException { /* nothing */ }
+
+
+  public InputStream wrapLayerDecoder(InputStream jis, LayerRuntime lr) throws IOException { return null; }
+
+  public OutputStream wrapLayerEncoder(OutputStream jos, LayerRuntime lr) throws IOException { return null; }
```

Comment:
Thoughts on naming these both `wrap` with either `InputStream` or `OutputStream` parameter types? Limiting layers don't really encode/decode so it's maybe confusing?